app/application/services/epoch_service.py

```
@@ -0,0 +1,110 @@
+        consistently. This method is for read-only resolution or bulk flows that call
+        record_events_appended_bulk after persisting.
+        """
+        existing = await self._epoch_repo.get_open_epoch_for_update(
+            tenant_id, subject_id
+        )
+        if existing is not None:
+            return existing
+
+        tenant = await self._tenant_repo.get_by_id(tenant_id)
+        if not tenant:
+            raise ValueError(f"Tenant not found: {tenant_id}")
```

Ruff (0.15.5)

[warning] Avoid specifying long messages outside the exception class

(TRY003)